### PR TITLE
String dtype: implement _get_common_dtype

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
This PR implements `StringDtype._get_common_dtype` to control the behaviour in concat etc.

This fixes two things:
- When concatting with a numpy string-like dtype (S/U), we can preserve our StringDtype instead of converting to object dtype (as happens now). In general we don't store such numpy dtypes in pandas objects, but in certain corner cases we are passing numpy dtypes directly to `find_common_type`.
- When concatting different instances of StringDtype. In practice this shouldn't happen often, but can certainly occur. In this case, I decided to give precedence to the pyarrow / pd.NA: 1) if pyarrow is installed, that's the default and therefore logical pick in this case as well, and 2) pd.NA is opt in, so if you have a mix, assume the user explicitly opted in and propagate that preference.

xref https://github.com/pandas-dev/pandas/issues/54792

### CODE REVIEW

File: pandas/core/arrays/string_.py [hunk 1]
```diff
@@ -284,6 +284,34 @@ def construct_array_type(  # type: ignore[override]
...
 
+    def _get_common_dtype(self, dtypes: list[DtypeObj]) -> DtypeObj | None:
+        storages = set()
+        na_values = set()
+
+        for dtype in dtypes:
+            if isinstance(dtype, StringDtype):
+                storages.add(dtype.storage)
+                na_values.add(dtype.na_value)
+            elif isinstance(dtype, np.dtype) and dtype.kind == "U":
```
Comment (mroeschke) [inline pandas/core/arrays/string_.py:295]: Would this handle the new numpy string type?

File: pandas/core/arrays/string_.py [hunk 2]
```diff
@@ -284,6 +284,34 @@ def construct_array_type(  # type: ignore[override]
...
+                continue
+            else:
+                return None
+
+        if len(storages) == 2:
+            # if both python and pyarrow storage -> priority to pyarrow
+            storage = "pyarrow"
+        else:
+            storage = next(iter(storages))  # type: ignore[assignment]
```
Comment (mroeschke) [inline pandas/core/arrays/string_.py:304]: Would it possible for `_get_common_dtype` to get all numpy string-like types where `storages` ends up empty after the `dtypes` loop?

File: pandas/core/arrays/string_.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py]: Expanded the `kind` check to let it also support numpy 2.0 strings

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jorisvandenbossche

**Pull request merged**